doc: add install guide with ODH operator (#57)
opendatahub-io/llm-d-batch-gateway-operator: docs/odh-installation.md

# ODH Installation Guide

Deploy LLM Batch Gateway through the ODH platform on OpenShift.

## Architecture

```
opendatahub-operator (platform orchestrator)
  → ai-gateway-operator (module operator)

>     → batch-gateway-operator (sub-module operator)

Reviewer: ```suggestion
    → llm-dbatch-gateway-operator (sub-module operator)
```

Author: fixed, thans

      → operands: apiserver, processor, gc
```

1. **User** creates a DataScienceCluster CR with `aigateway.managementState: Managed` and `aigateway.batchGateway.managementState: Managed`
2. **ODH operator** renders ai-gateway-operator's kustomize manifests and deploys it (Deployment, RBAC, CRD) into the applications namespace
3. **ODH operator** creates an AIGateway CR with the batch gateway config projected from DSC

> 4. **ai-gateway-operator** reconciles the AIGateway CR and deploys batch-gateway-operator (Deployment, RBAC, CRD)

Reviewer: ```suggestion
4. **ai-gateway-operator** reconciles the AIGateway CR and deploys llm-dbatch-gateway-operator (Deployment, RBAC, CRD)
```

Author: fixed, thanks

5. **User** creates an LLMBatchGateway CR in a workload namespace with database, storage, and inference gateway config

> 6. **batch-gateway-operator** reconciles the LLMBatchGateway CR and deploys the operands: API server, request processor, and garbage collector

Reviewer: ```suggestion
6. **llm-dbatch-gateway-operator** reconciles the LLMBatchGateway CR and deploys the operands: API server, request processor, and garbage collector
```

Author: fixed, thanks

## Prerequisites

- OpenShift cluster with ODH operator installed
- `oc` / `kubectl` CLI connected to the cluster
- `helm` (for installing prerequisites: PostgreSQL, Redis, MinIO)
- External dependencies for LLMBatchGateway: PostgreSQL, Redis, S3-compatible storage

## Installation

### Step 1: Create DSCI and DSC with Batch Gateway enabled

```bash
# DSCInitialization
oc apply -f - <<'EOF'
apiVersion: dscinitialization.opendatahub.io/v2
kind: DSCInitialization
metadata:
  name: default-dsci
spec:
  applicationsNamespace: opendatahub
  monitoring:
    managementState: Removed
    namespace: opendatahub
  trustedCABundle:
    managementState: Removed
EOF

# DataScienceCluster with aigateway + batchGateway enabled
oc apply -f - <<'EOF'
apiVersion: datasciencecluster.opendatahub.io/v2
kind: DataScienceCluster
metadata:
  name: default-dsc
spec:
  components:
    aigateway:
      managementState: Managed
      batchGateway:
        managementState: Managed
EOF
```

This triggers the deployment chain:
1. ODH operator deploys ai-gateway-operator in `opendatahub` namespace
2. ODH operator creates AIGateway CR (cluster-scoped singleton)

> 3. ai-gateway-operator deploys batch-gateway-operator

Reviewer: ```suggestion
3. ai-gateway-operator deploys llm-d-batch-gateway-operator
```

Author: fixed, thanks

> 4. Both operators become Ready

Reviewer: maybe remove this? just keep the checks below

Author: fixed, thanks

Check AI Gateway status:

```bash
$ oc get aigateway
NAME                READY   REASON   VERSION
default-aigateway   True             0.0.0-dev

$ oc get aigateway default-aigateway -o yaml
apiVersion: components.platform.opendatahub.io/v1alpha1
kind: AIGateway
metadata:
  name: default-aigateway
  labels:
    platform.opendatahub.io/part-of: datasciencecluster
  ownerReferences:
  - apiVersion: datasciencecluster.opendatahub.io/v2
    kind: DataScienceCluster
    name: default-dsc
spec:
  batchGateway:
    managementState: Managed
status:
  conditions:
  - status: "True"
    type: Ready
  - status: "True"
    type: ProvisioningSucceeded
  - status: "True"
    type: DeploymentsAvailable
  phase: Ready

$ oc get deployments -n opendatahub
NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
ai-gateway-operator      1/1     1            1           2m

> batch-gateway-operator   1/1     1            1           90s

Reviewer: ```suggestion
llm-d-batch-gateway-operator   1/1     1            1           90s
```

Author: I see, my fork is out-of-sync, still using old name `batch-gateway-operator`
thanks for pointing out!

Author: fixed, thanks

```

### Step 2: Set up external dependencies

LLMBatchGateway requires PostgreSQL, Redis, and S3-compatible storage. For testing, use the provided script:

```bash

> NAMESPACE=batch-api bash hack/setup-prereqs.sh

Reviewer: should we just `export NAMESPACE=batch-api`
and then refer to NAMESPACE below than hardcoded ?

Author: fixed, thanks

      requestTimeout: 5m
      maxRetries: 3
      initialBackoff: 1s
      maxBackoff: 60s
  gc:
    interval: 30m
EOF
```

### Step 4: Check Batch Gateway status

```bash
$ oc get llmbatchgateway -n batch-api
NAMESPACE   NAME            READY   API-READY   PROC-READY   GC-READY   AGE
batch-api   batch-gateway   True    1           1            1          20s

$ oc get pods -n batch-api
NAME                                       READY   STATUS    RESTARTS   AGE
batch-gateway-apiserver-5967fbb99b-nx22h   1/1     Running   0          25s
batch-gateway-gc-787ccdd97f-lqgk2          1/1     Running   0          25s
batch-gateway-processor-7dfc4579c7-vl59r   1/1     Running   0          25s

$ oc get llmbatchgateway batch-gateway -n batch-api -o yaml
apiVersion: batch.llm-d.ai/v1alpha1
kind: LLMBatchGateway
metadata:
  name: batch-gateway
  namespace: batch-api
spec:
  apiServer:
    replicas: 1
  dbBackend: postgresql
  fileStorage:
    s3:
      accessKeyId: minioadmin
      autoCreateBucket: true
      endpoint: http://minio.batch-api.svc.cluster.local:9000
      region: us-east-1
      usePathStyle: true
  gc:
    interval: 30m
  processor:
    globalInferenceGateway:

>       url: http://vllm-sim.default.svc.cluster.local:8000

Reviewer 2: The sample output here shows `vllm-sim.default.svc.cluster.local` but the CR in Step 3 uses `inference-gateway.default.svc.cluster.local`. These should match.

Author: fixed, thanks

      requestTimeout: 5m
      maxRetries: 3
      initialBackoff: 1s
      maxBackoff: 60s
    replicas: 1
  secretRef:
    name: batch-gateway-secrets
status:
  componentStatus:
    apiServer:
      readyReplicas: 1
      replicas: 1
    gc:
      readyReplicas: 1
      replicas: 1
    processor:
      readyReplicas: 1
      replicas: 1
  conditions:
  - message: API server has at least one ready replica
    reason: Available
    status: "True"
    type: APIServerAvailable
  - message: Processor has at least one ready replica
    reason: Available
    status: "True"
    type: ProcessorAvailable
  - message: GC has at least one ready replica
    reason: Available
    status: "True"
    type: GCAvailable
  - message: All components have at least one ready replica
    reason: AllComponentsReady
    status: "True"
    type: Ready
```

## Disable / Remove

```bash
# Disable AI Gateway (removes operators in reverse order)
oc patch datasciencecluster default-dsc --type merge \

>     -p '{"spec":{"components":{"aigateway":{"managementState":"Removed","batchGateway":{"managementState":"Removed"}}}}}'

Reviewer: ```suggestion
    -p '{"spec":{"components":{"aigateway":{"managementState":"Removed"}}}}'
```

set aigateway to Removed should be enough to delete AIGateway CR and aggrdated with all

Author: fixed, thanks
